### PR TITLE
feat(docs): add GitHub Handler description

### PR DESCRIPTION
**Pull request by k-dovgan**

### CODE REVIEW

**Comment by k-dovgan** [comment]
https://universum.readthedocs.io/en/add-description/args_github_handler.html

File: doc/args_github_handler.rst [hunk 1]
```diff
@@ -3,7 +3,18 @@
...
+builds to perform these checks, parsing accepted payload for required params and passing them to usual
+Universum runs in these builds.
+
+For GitHub Handler to work, it needs:
+
+* the payload, sent by GitHub in web-hook to GitHub Application, to be passed to ``--payload`` parameter
+* event name, sent by GitHub in x-github-event" web-hook header, to be passed to ``--event`` parameter
+* an URL to trigger check run build to be passed to ``--trigger-url`` parameter
+
+Additional parameters are described below.
```
Comment (i-keliukh) [inline doc/args_github_handler.rst:17]: The parameters mentioned above are also listed in the doc. Therefore, these are not additional parameters.

File: doc/args_github_handler.rst [hunk 2]
```diff
@@ -3,7 +3,18 @@
...
-.. Insert some description here
+GitHub Handler is Universum mode that serves as GitHub Application, helping to perform and report checks on
+new commits to a repository. In this mode Universum can create new check runs on GitHub and trigger
+builds to perform these checks, parsing accepted payload for required params and passing them to usual
+Universum runs in these builds.
+
+For GitHub Handler to work, it needs:
+
+* the payload, sent by GitHub in web-hook to GitHub Application, to be passed to ``--payload`` parameter
+* event name, sent by GitHub in x-github-event" web-hook header, to be passed to ``--event`` parameter
```
Comment (i-keliukh) [inline doc/args_github_handler.rst:14]: Strange quote (") character

File: doc/args_github_handler.rst [hunk 3]
```diff
@@ -3,7 +3,45 @@
...
+GitHub Handler is Universum mode that serves as GitHub Application, helping to perform and report checks on
+new commits to a repository. In this mode Universum can create new check runs on GitHub and trigger
+builds to perform these checks, parsing accepted payload for required params and passing them to usual
+Universum runs in these builds. The following picture represents the event sequence leading to check result
+for a commit displayed on GitHub:
+
+*(picture here)*
+
+On this picture numbers depict the following events:
+
```
Comment (i-keliukh) [inline doc/args_github_handler.rst:15]: Please either add a short description of what has to be setup to make the described workflow to work, or change the workflow description into the setup tutorial.

File: doc/args_github_handler.rst [hunk 4]
```diff
@@ -3,7 +3,45 @@
...
+1. User makes a new commit
+2. GitHub detects a commit, creates a check suite and sends a web-hook payload to all subscribed entities
```
Comment (i-keliukh) [inline doc/args_github_handler.rst:17]: It should be enough to mention that github sends the payload to our handler.

File: doc/args_github_handler.rst [hunk 5]
```diff
@@ -3,7 +3,45 @@
...
+   (including GitHub Handler)
+3. GitHub Handler checks that conditions (repo name, web-hook event) are applicable and sends to GitHub a
```
Comment (i-keliukh) [inline doc/args_github_handler.rst:19]: Remove the word "conditions" and describe those directly

File: doc/args_github_handler.rst [hunk 6]
```diff
@@ -3,7 +3,45 @@
...
+   request to create a new check run
+4. GitHub creates a new check run and sends a new web-hook payload about this event
+5. GitHub Handler triggers a usual `Universum` run, retrieving all required parameters from GitHub web-hook
```
Comment (i-keliukh) [inline doc/args_github_handler.rst:22]: Definitely automation server has to be mentioned here

File: doc/args_github_handler.rst [hunk 7]
```diff
@@ -3,7 +3,45 @@
...
+
+*(picture here)*
+
+On this picture numbers depict the following events:
+
+1. User makes a new commit
+2. GitHub detects a commit, creates a check suite and sends a web-hook payload to all subscribed entities
+   (including GitHub Handler)
+3. GitHub Handler checks that conditions (repo name, web-hook event) are applicable and sends to GitHub a
+   request to create a new check run
```
Comment (i-keliukh) [inline doc/args_github_handler.rst:20]: change to "...and sends a request to create a new check run to GitHub"

File: doc/github_handler.rst [hunk 8]
```diff
@@ -0,0 +1,60 @@
+GitHub Handler
+==============
+
+GitHub Handler is a Universum mode that serves as GitHub Application, helping to perform and report checks on
+new commits to a repository. In this mode Universum can create new check runs on GitHub and trigger an already
+set up automation server to perform these checks. GitHub Handler parses all required params and passes them
```
Comment (i-keliukh) [inline doc/github_handler.rst:6]: We should clearly separate main mode and handler mode in the text. In this and previous line two different names for the same thing are used: "this Universum mode" and "GitHub Handler". While technically this is correct, it is confusing for the reader.

I propose to either never call handler "a mode" or always refer to it as "GitHub handler mode".

File: doc/github_handler.rst [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+to the triggered builds.
+
+
+How to set up GitHub auto check using Unversum
+----------------------------------------------
+
+Universum in default, 'main' mode when set up to ``--vcs-type=github`` and with ``--report-to-review`` passed
```
Comment (i-keliukh) [inline doc/github_handler.rst:13]: This sentence is not grammatically correct.
I think we can split it and make more clear.

File: doc/github_handler.rst [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+as parameter, can post `check run` statuses to GitHub to be depicted on 'Checks' page in pull requests or as
+a simple icon near any checked commit. To do so, it needs `a set of parameters <args.html#GitHub>`__
+to be passed to it.
+
+.. note::
+
+    During these checks, if some steps were using :doc:`Universum analyzers <code_report>`,
```
Comment (i-keliukh) [inline doc/github_handler.rst:20]: Can we add link to the example?
May be create some permanent demo branch in the repository?

File: doc/github_handler.rst [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    and any issues are found, they will also be reported in comments.
+
+These required parameters are sent by GitHub via web-hooks after the reuired `check run` is created. To do so,
```
Comment (i-keliukh) [inline doc/github_handler.rst:23]: @k-dovgan, could you, please confirm all of these parameters are provided when using GitHub Handler?
If yes, please change the wording to emphasize this.
If no, please provide the exact list of parameters.

File: doc/github_handler.rst [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+a GitHub Application is required, and GitHub Handler could be used as such.
+
+.. note::
+
+    GitHub Handler is a simple script, receiving web-hook payload and event as parameters.
+    To work as GitHub Application it still needs a server to receive actual web-hooks and pass their
```
Comment (i-keliukh) [inline doc/github_handler.rst:29]: server -> web server

File: doc/github_handler.rst [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    contents as parameters to Handler.
+
+One of the easiest ways to make it work is to put the GitHub Handler execution into an automation server
+(such as Jenkins) job triggered by incoming web-hooks from GitHub. GitHub Handler parses the incoming
```
Comment (i-keliukh) [inline doc/github_handler.rst:33]: Does Jenkins require some plugin for this to work?
Also, what do you think about providing exact instructions for setting up Jenkins?

File: doc/github_handler.rst [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+web-hook event (passed via "x-github-event" header) and payload content. If a new `check run` is required,
+it sends a request to GitHub to create one. If the `check run` is already created, it parses the required parameters
+and triggers the set up automation server to run a previously described check by usual Universum.
```
Comment (i-keliukh) [inline doc/github_handler.rst:36]: In my opinion, the reference to "usual Universum" doesn't help here.
It must be at least mentioned in advance.

File: doc/github_handler.rst [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+The following picture represents the event sequence leading to check result for a commit displayed on GitHub:
+
+*(picture here)*
```
Comment (i-keliukh) [inline doc/github_handler.rst:40]: We need to finish the picture

File: doc/github_handler.rst [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+On this picture numbers depict the following events:
+
+1. User makes a new commit
+2. GitHub detects a commit, creates a check suite and sends a web-hook payload to GitHub Handler
```
Comment (i-keliukh) [inline doc/github_handler.rst:45]: To Jenkins, not directly to the Handler

File: doc/github_handler.rst (hunk 15 above)
Comment (i-keliukh) [reply doc/github_handler.rst:53]: I propose the following picture:
githubhandler_plain

There is a vector (svg) version.

File: doc/github_handler.rst [hunk 17]
```diff
@@ -58,23 +58,36 @@ On this picture numbers depict the following events:
 
 1. User makes a new commit
 2. GitHub detects a commit, creates a check suite and sends a web-hook payload to Jenkins server
-3. A preconfigured Jenkins job passes received parameters to GitHub Handler
-4. GitHub Handler checks that repo name and web-hook event are applicable and and sends a request
+3. `Generic Webhook Trigger` plugin triggers a preconfigured Jenkins job ('GitHub Webhook handler') on an already set up
+   automation server, passing required parameters, such as payload contents and "x-github-event" header
```
Comment (i-keliukh) [inline doc/github_handler.rst:62]: 1. You already mentioned Jenkins server in item 2, but now you refer to "automation server".
Please use Jenkins for consistency.
2. What form is used to pass parameter?
I think this sentence will look better if you mention it explicitly.

File: doc/github_handler.rst [hunk 18]
```diff
@@ -58,23 +58,36 @@ On this picture numbers depict the following events:
...
+4. GitHub Handler checks that repo name and web-hook event are applicable, and if so, sends a request
```
Comment (i-keliukh) [inline doc/github_handler.rst:63]: I think you can mention that GitHub Handler uses GitHub API for creating check runs.

File: doc/github_handler.rst (hunk 17 above)
Comment (i-keliukh) [reply doc/github_handler.rst:62]: 3. You can also mention what is x-github-event

File: doc/github_handler.rst [hunk 19]
```diff
@@ -58,23 +58,36 @@ On this picture numbers depict the following events:
...
    to create a new check run to GitHub
 5. GitHub creates a new check run and sends a new web-hook payload about this event
-6. GitHub Handler triggers a usual `Universum` run on an already set up automation server,
-   retrieving all required parameters from GitHub web-hook payload and passing them to the build
-7. Universum in default mode with ``--report-to-review`` on and ``--vcs-type=github`` uses received parameters
-   to report build ('check') start
-8. Depending on build result, Universum either reports build success or failure
-
-After this, check result can be viewed directly on GitHub.
+6. `Generic Webhook Trigger` plugin triggers the 'GitHub Webhook handler' job again, now with different event and
```
Comment (i-keliukh) [inline doc/github_handler.rst:66]: "event type"?
Please mention it is x-github-event header.
Also, please specify its value exactly.

File: doc/github_handler.rst [hunk 20]
```diff
@@ -58,23 +58,36 @@ On this picture numbers depict the following events:
...
+   payload
+7. Based on received event, GitHub Handler triggers another preconfigured job with usual `Universum` call
+   ('Check commit'), retrieving all required parameters from GitHub web-hook payload and passing them to the build
+   via `build with parameters` trigger mechanism
+
+   * this job may be configured on another Jenkins instance
```
Comment (i-keliukh) [inline doc/github_handler.rst:72]: I think you don't need to mention another Jenkins here.

File: doc/github_handler.rst [hunk 21]
```diff
@@ -58,23 +58,36 @@ On this picture numbers depict the following events:
...
-6. GitHub Handler triggers a usual `Universum` run on an already set up automation server,
-   retrieving all required parameters from GitHub web-hook payload and passing them to the build
-7. Universum in default mode with ``--report-to-review`` on and ``--vcs-type=github`` uses received parameters
-   to report build ('check') start
-8. Depending on build result, Universum either reports build success or failure
-
-After this, check result can be viewed directly on GitHub.
+6. `Generic Webhook Trigger` plugin triggers the 'GitHub Webhook handler' job again, now with different event and
+   payload
+7. Based on received event, GitHub Handler triggers another preconfigured job with usual `Universum` call
```
Comment (i-keliukh) [inline doc/github_handler.rst:68]: In my opinion it is better to keep details regarding 'Check commit' and universum in item 8.
Please also mention that the trigger is done using URL (http request).

File: doc/github_handler.rst [hunk 22]
```diff
@@ -58,23 +58,36 @@ On this picture numbers depict the following events:
...
+   ('Check commit'), retrieving all required parameters from GitHub web-hook payload and passing them to the build
+   via `build with parameters` trigger mechanism
+
+   * this job may be configured on another Jenkins instance
+8. Receiving all required parameters, Universum in default mode with ``--report-to-review`` on and ``--vcs-type=github``
+   performs a commit check according to :doc:`configuration file <configuring>`,
+   set up in `build parameters <args.html#Configuration\ execution>`__
+9. [optional] If ``--report-build-start`` set in `build parameters <args.html#Result\ reporting>`__, GitHub will be
```
Comment (i-keliukh) [inline doc/github_handler.rst:76]: Please change to active voice ("Universum reports check is already in progress").

File: doc/github_handler.rst (hunk 20 above)
Comment (k-dovgan) [reply doc/github_handler.rst:72]: But it is important, and not obvious at all, isn't it?

Comment (i-keliukh) [reply doc/github_handler.rst]: You can mention it below the workflow in a side note.

File: doc/github_handler.rst [hunk 23]
```diff
@@ -57,24 +57,28 @@ The following picture represents a possible event sequence leading to a check re
...
 
 1. User makes a new commit
-2. GitHub detects a commit, creates a check suite and sends a web-hook payload to Jenkins server
-3. `Generic Webhook Trigger` plugin triggers a preconfigured Jenkins job ('GitHub Webhook handler') on an already set up
-   automation server, passing required parameters, such as payload contents and "x-github-event" header
-4. GitHub Handler checks that repo name and web-hook event are applicable, and if so, sends a request
+2. GitHub detects a commit, creates a check suite and sends a web-hook payload with event ``check_suite``
+   in "x-github-event" header to an already set up automation server (in this case, Jenkins)
+3. `Generic Webhook Trigger` plugin triggers a preconfigured Jenkins job ('GitHub Webhook handler'),
+   passing payload contents via ``genericVariable`` and "x-github-event" header via ``genericHeaderVariable``
```
Comment (i-keliukh) [inline doc/github_handler.rst:63]: What are they?
Environment variables?
Job parameters?

File: doc/github_handler.rst [hunk 24]
```diff
@@ -57,24 +57,28 @@ The following picture represents a possible event sequence leading to a check re
...
+4. GitHub Handler checks that repo name and web-hook event are applicable, and if so, sends a request via
+   `GitHub API <https://docs.github.com/en/rest/reference/checks#create-a-check-suite>`__
    to create a new check run to GitHub
-5. GitHub creates a new check run and sends a new web-hook payload about this event
+5. GitHub creates a new check run and sends a new web-hook payload with event ``check_run``
```
Comment (i-keliukh) [inline doc/github_handler.rst:67]: "with event ``check_run``" -> "with ``check_run`` event"

File: doc/github_handler.rst [hunk 25]
```diff
@@ -57,24 +57,28 @@ The following picture represents a possible event sequence leading to a check re
 On this picture numbers depict the following events:
 
 1. User makes a new commit
-2. GitHub detects a commit, creates a check suite and sends a web-hook payload to Jenkins server
-3. `Generic Webhook Trigger` plugin triggers a preconfigured Jenkins job ('GitHub Webhook handler') on an already set up
-   automation server, passing required parameters, such as payload contents and "x-github-event" header
-4. GitHub Handler checks that repo name and web-hook event are applicable, and if so, sends a request
+2. GitHub detects a commit, creates a check suite and sends a web-hook payload with event ``check_suite``
```
Comment (i-keliukh) [inline doc/github_handler.rst:60]: "with event ``check_suite``" -> "with ``check_suite`` event"

**Review by o-andrieiev — APPROVED** [review]

**Review by i-keliukh — APPROVED** [review]

**Pull request merged**